rust/src/id.rs

```diff
@@ -0,0 +1,128 @@
+use std::str::FromStr;
+
+use percent_encoding::{percent_decode, percent_encode, NON_ALPHANUMERIC};
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct FullId {
+    pub scheme: String,
+    pub id: String,
+    pub path: String,
+}
+
+impl FullId {
+    pub fn new(scheme: String, id: String, path: String) -> Self {
+        FullId { scheme, id, path }
+    }
+
+    pub fn split(self) -> (BackendId, String) {
+        return (BackendId::new(self.scheme, self.id), self.path);
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct BackendId {
+    pub scheme: String,
+    pub id: String,
+}
+
+impl BackendId {
+    pub fn new(scheme: String, id: String) -> Self {
+        BackendId { scheme, id }
+    }
+
+    pub fn relative(&self, path: String) -> FullId {
+        FullId::new(self.scheme.clone(), self.id.clone(), path)
+    }
+
+    pub fn user_to_relative(&self, user: crate::proto::User) -> crate::proto::User {
+        crate::proto::User {
+            id: self.relative(user.id).to_string(),
+            display_name: user.display_name,
+        }
+    }
+
+    pub fn source_to_relative(
+        &self,
+        source: crate::proto::ChannelSource,
+    ) -> crate::proto::ChannelSource {
+        crate::proto::ChannelSource {
+            channel_id: self.relative(source.channel_id).to_string(),
+            user: source.user.map(|user| self.user_to_relative(user)),
+        }
+    }
+}
+
+impl Into<BackendId> for FullId {
+    fn into(self) -> BackendId {
+        BackendId {
+            scheme: self.scheme,
+            id: self.path,
+        }
+    }
```

Comment:
Same here, do you not need to use the `FullId` after you've `.into()`d?

Reply:
Usually `into` is used when you're taking ownership and turning it into something, `to_*` is used when it's being done as a reference to the parent object. I actually dropped this impl because it's never used.

Reply:
Into typically doesn’t take ownership I thought